### PR TITLE
[Storage] Support high throughput append blob

### PR DESCRIPTION
**Pull request by yifanz7**
## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Storage/Storage/Blob/Cmdlet/SetAzureStorageBlobContent.cs [hunk 1]
```diff
@@ -542,32 +544,40 @@ internal virtual async Task UploadBlobwithSdk(long taskId, IStorageBlobManagemen
...
+                        }
+                        byte[] uploadcacheLong = null;
                         byte[] uploadcache = null;
                         progressHandler.Report(0);
                         long offset = 0;
                         while (offset < fileSize)
                         {
                             // Get chunk size and prepare cache
-                            int chunksize = size4MB;
+                            //int chunksize = size4MB;
```
Comment (blueww) [inline src/Storage/Storage/Blob/Cmdlet/SetAzureStorageBlobContent.cs:562]: remove this line

**Review by blueww — CHANGES_REQUESTED** [review]
Would you please also indicate which tests you have done with the change, like include which blob size? which kind of pipeline? with -asjob or not.

**Comment by yifanz7** [comment]
> Would you please also indicate which tests you have done with the change, like include which blob size? which kind of pipeline? with -asjob or not.

Tested the following cases with maxBlobCount = 10 and DMLib branch commented out, in order to cover all scenarios of the chunk size: 

- Upload blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob). 
- Copy blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob)
- Copy blobs of size 5MB and 333MB by piping in the blob object 

Tested the following case with maxBlobCount = 1 and DMLib branch commented out: 

- Upload a blob of size 333MB. This is expected to fail

**Comment by blueww** [comment]
> > Would you please also indicate which tests you have done with the change, like include which blob size? which kind of pipeline? with -asjob or not.
> 
> Tested the following cases with maxBlobCount = 10 and DMLib branch commented out, in order to cover all scenarios of the chunk size:
> 
> * Upload blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob).
> * Copy blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob)
> * Copy blobs of size 5MB and 333MB by piping in the blob object
> 
> Tested the following case with maxBlobCount = 1 and DMLib branch commented out:
> 
> * Upload a blob of size 333MB. This is expected to fail

Thanks @yifanz7!
I assume the blob type in all above test are append blob.
Besides that, we might should also change the default max block count to a smaller size (like 3), and run the test to see the >8MB block can be uploaded/copied.

**Comment by yifanz7** [comment]
> > > Would you please also indicate which tests you have done with the change, like include which blob size? which kind of pipeline? with -asjob or not.
> > 
> > 
> > Tested the following cases with maxBlobCount = 10 and DMLib branch commented out, in order to cover all scenarios of the chunk size:
> > 
> > * Upload blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob).
> > * Copy blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob)
> > * Copy blobs of size 5MB and 333MB by piping in the blob object
> > 
> > Tested the following case with maxBlobCount = 1 and DMLib branch commented out:
> > 
> > * Upload a blob of size 333MB. This is expected to fail
> 
> Thanks @yifanz7! I assume the blob type in all above test are append blob. Besides that, we might should also change the default max block count to a smaller size (like 3), and run the test to see the >8MB block can be uploaded/copied.

Yes they're all append blobs. 
For block of size > 8MB, I think it's covered with size 200MB and 333MB for max block count = 10. 
For 333MB blob, the block size will be (333/10 + 1) * 4 = 136MB 
Please correct me if I'm wrong.

**Comment by yifanz7** [comment]
> > > Would you please also indicate which tests you have done with the change, like include which blob size? which kind of pipeline? with -asjob or not.
> > 
> > 
> > Tested the following cases with maxBlobCount = 10 and DMLib branch commented out, in order to cover all scenarios of the chunk size:
> > 
> > * Upload blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob).
> > * Copy blobs of size 3MB, 5MB, 10MB, 200MB and 333MB (with -AsJob and without -AsJob)
> > * Copy blobs of size 5MB and 333MB by piping in the blob object
> > 
> > Tested the following case with maxBlobCount = 1 and DMLib branch commented out:
> > 
> > * Upload a blob of size 333MB. This is expected to fail
> 
> Thanks @yifanz7! I assume the blob type in all above test are append blob. Besides that, we might should also change the default max block count to a smaller size (like 3), and run the test to see the >8MB block can be uploaded/copied.

Also have tested uploading with relative paths, both with AsJob and without AsJob. It worked well

**Review by blueww — APPROVED** [review]

**Pull request merged**